embassy-sync/src/mutex.rs

```diff
@@ -180,3 +198,122 @@ where
         unsafe { &mut *(self.mutex.inner.get()) }
     }
 }
+
+/// A handle to a held `Mutex` that has had a function applied to it via [`MutexGuard::map`] or
+/// [`MappedMutexGuard::map`].
+///
+/// This can be used to hold a subfield of the protected data.
+#[clippy::has_significant_drop]
+pub struct MappedMutexGuard<'a, M, T, U>
+where
+    M: RawMutex,
+    T: ?Sized,
+    U: ?Sized,
+{
+    mutex: &'a Mutex<M, T>,
+    value: *mut U,
+}
+
+impl<'a, M, T, U> MappedMutexGuard<'a, M, T, U>
+where
+    M: RawMutex,
+    T: ?Sized,
+    U: ?Sized,
+{
+    /// Returns a locked view over a portion of the locked data.
+    pub fn map<V>(this: Self, fun: impl FnOnce(&mut U) -> &mut V) -> MappedMutexGuard<'a, M, T, V> {
+        let mutex = this.mutex;
+        let value = fun(unsafe { &mut *this.value });
+        // Don't run the `drop` method for MutexGuard. The ownership of the underlying
+        // locked state is being moved to the returned MappedMutexGuard.
+        mem::forget(this);
+        MappedMutexGuard { mutex, value }
+    }
+}
+
+impl<'a, M, T, U> Drop for MappedMutexGuard<'a, M, T, U>
+where
+    M: RawMutex,
+    T: ?Sized,
+    U: ?Sized,
+{
+    fn drop(&mut self) {
+        self.mutex.state.lock(|s| {
+            let mut s = unwrap!(s.try_borrow_mut());
+            s.locked = false;
+            s.waker.wake();
+        })
+    }
+}
+
+impl<'a, M, T, U> Deref for MappedMutexGuard<'a, M, T, U>
+where
+    M: RawMutex,
+    T: ?Sized,
+    U: ?Sized,
+{
+    type Target = U;
+    fn deref(&self) -> &Self::Target {
+        // Safety: the MutexGuard represents exclusive access to the contents
+        // of the mutex, so it's OK to get it.
+        unsafe { &*self.value }
+    }
+}
+
+impl<'a, M, T, U> DerefMut for MappedMutexGuard<'a, M, T, U>
+where
+    M: RawMutex,
+    T: ?Sized,
+    U: ?Sized,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        // Safety: the MutexGuard represents exclusive access to the contents
+        // of the mutex, so it's OK to get it.
+        unsafe { &mut *self.value }
+    }
+}
+
+unsafe impl<M, T, U> Send for MappedMutexGuard<'_, M, T, U>
```

Comment:
implementing Send/Sync is only sound if `M: Sync`. 

I think if you make MappedMutexGuard contain `&mut U` instead of `*mut U` Rust will be able to automatically infer the right Send/Sync bounds, so no `unsafe` needed.

(embassy-sync has a noop RawMutex that's not Sync, which allows creating async mutexes that can only be shared between tasks in the same thread, not between threads. this is somewhat unique vs other async runtimes)

Reply:
Bounds have been added to `M`.

All implementations I saw use pointers so I am not sure if references are possible.